### ISSUE TITLE
PerformanceResourceTiming: startTime

### ISSUE THREAD

**Issue by avadacatavra** [servo/servo#21254]
In components/net/http_loader.rs::http_fetch we need to set start_time for the PerformanceEntry member of PerformanceResourceTiming. This will be equal to either fetch_start or redirect_start.

Spec: https://w3c.github.io/resource-timing/#dfn-starttime

**Comment by chirgjn** [comment]
Hey @avadacatavra 👋
Came here from your tweet, would love to take this up. But I'm fairly new to rust, I've got plenty of time this week though. Should I assign it to myself?

**Comment by chirgjn** [comment]
@highfive: assign me

Taking it up for now :D

**Comment by avadacatavra** [comment]
Hi @chirgjn--thanks for taking this on!

Let me know if there's any help you need

**Comment by chirgjn** [comment]
Hey, thanks. I'm able to build servo on my machine but apart from that I'm having a hard time.
I've read through the comments and the spec, and what I understand is that first need to find out the fetch_start and redirect_start right?
Can you point me where to look?

**Comment by jdm** [comment]
@chirgjn Since this attribute relies on some other attributes that aren't implemented yet (like #21258) it probably makes more sense to start with those ones.

**Comment by aditj** [comment]
@jdm Since [FetchStart](https://w3c.github.io/resource-timing/#dom-performanceresourcetiming-fetchstart) is now [set up](https://github.com/servo/servo/issues/21258), Can I work on this? :smile: 
(Note:  startTime can be set to either FetchStart or RedirectStart as mentioned in the specs).

**Comment by aditj** [comment]
@chirgjn :Thanks :smile: and you can take it up if it can actually be done(i.e. to say there are no other dependencies left :P) 
Also i dont have a problem in collaborating but I dont know how that is gonna work :/

**Comment by chirgjn** [comment]
@aditj I'll take it up then.
This is currently blocked on #21256 I think which will be resolved after https://github.com/servo/servo/pull/22319 gets merged after that I don't think there are any blockers.

**Comment by jdm** [comment]
I suggest that we add an API to ResourceFetchTiming called set_attribute_from that takes two ResourceAttribute arguments and sets the first attribute based on the value of the second attribute.

**Comment by aditj** [comment]
@jdm should we file a seperate issue for that ? 
And if no can I create a new pull request with this API? 
(and @chirgjn can maybe implement it for startTime)

**Comment by jdm** [comment]
I was thinking it could be added as part of fixing this issue.

**Comment by davidweitzenfeld** [comment]
@highfive: assign me

**Comment by jdm** [comment]
@davidweitzenfeld It looks like you made a pull request against your fork, not servo/servo :)

**Comment by jdm** [comment]
@davidweitzenfeld It looks like you made progress but never returned to make the PR against the right repository. Are you still planning to complete this work?

**Comment by jdm** [comment]
Unassigning due to lack of response.

**Comment by pag4k** [comment]
Hello! This sounds like a perfect first issue for me. I some Rust experience, but I never contributed to such a project.

If it is possible, then: @highfive: assign me :-)

**Comment by pag4k** [comment]
Hello!
I have a two small questions about this part of the specification: `If there are HTTP redirects or equivalent when fetching the resource, and if the timing allow check algorithm passes, this attribute MUST return the same value as redirectStart.  Otherwise, this attribute MUST return the same value as fetchStart.`

1. Would it be fine to first set startTime to fetchStart when the latter is set ([here](https://github.com/servo/servo/blob/31ee17e9e1c6012fd512f594c27abb44fe927ec4/components/net/fetch/methods.rs#L97)) and then only changing it to redirectStart later if all the above conditions are respected?

2. Speaking of the conditions, looking at [http_loader.rs](https://github.com/servo/servo/blob/31ee17e9e1c6012fd512f594c27abb44fe927ec4/components/net/http_loader.rs#L669), it seems that the `timing allow check` is not implemented yet. Is there something I am missing?

Thank you!

**Comment by jdm** [comment]
1. Yes, that sounds fine.
2. No, that's #21270.

**Comment by pag4k** [comment]
More questions!
This time about [performanceresourcetiming.rs](https://github.com/servo/servo/blob/master/components/script/dom/performanceresourcetiming.rs).
Two things that are probably related:
1. I'm not too sure how to deal with start_time in `new_inherited()`. Should it also be set to the fetch_start that is passed in argument? In a previous version of the [file](https://github.com/servo/servo/pull/22714/files?file-filters%5B%5D=.rs&owned-by%5B%5D=#diff-d1bf57f1536a05334bfaca1412110e2dL81), it used to be, but since it is not certain they will be the same value, I'm not sure.
2. It seems that the `//TODO fetch start should be in RFT` above `from_resource_timing()` is no longer relevant since we have `fetch_start: resource_timing.fetch_start as f64,` since [this](https://github.com/servo/servo/pull/22714/files?file-filters%5B%5D=.rs&owned-by%5B%5D=#diff-2de581f1e1bf3b6d63fd42bf3e8e93deR17).
Thank you!

**Comment by pag4k** [comment]
Well, I think I can answer my first question: `new_inherited()` is only used by [PerformanceNavigationTiming](https://github.com/servo/servo/blob/master/components/script/dom/performancenavigationtiming.rs) and [its specifications](https://w3c.github.io/navigation-timing/#performanceentry) says that `startTime attribute MUST return a DOMHighResTimeStamp with a time value of 0.` So based on that, startTime should be left to 0, but I'm not sure to what extend I can generalize based on the case PerformanceNavigationTiming...

**Comment by pajamapants3000** [comment]
@highfive: assign me (please!)

**Comment by pajamapants3000** [comment]
@jdm I created #24430 for adding the `set_attribute_from()` method you suggested, but I was wondering about a different possible approach. I was thinking of adding a `ResourceTimerValue` enum to, eventually, replace `RedirectStartValue` and `RedirectEndValue` and also be used for the future implementation of additional timers (like startTime). It would have variants `Zero`, `Now`, `FetchStart`, plus any additional timers whose values need to be copied. We could retain the existing API, just continuing to use `set_attribute`, and each of the variants of `ResourceAttribute` that refer to timer attributes could take a `ResourceTimerValue` to indicate how it is being set.
If that doesn't sound good, let me know and also I would appreciate some feedback on #24430 if you get a chance. Of course if the enum sounds better I'm happy to drop #24430. But in order to add startTime I need _some_ way of setting it! Thank you :)

**Comment by jdm** [comment]
I like that idea!

**Comment by pajamapants3000** [comment]
Hey folks, just an update: I didn't get lost; still working on this. I think I have a solution, I'm just figuring out the whole testing issue. This has proven rather time-consuming, given how long it takes to run the full test-wpt suite. That, and just familiarizing myself with everything.

I am currently attempting to see if my changes have resulted in any difference in the test-wpt results. It appears that most/all of the relevant wpt encompass multiple features, many of which (e.g. duration) have not yet been implemented. If implementing start time isn't enough to turn any FAILs to PASSes, I may add a unit-test or two and possibly some tests under tests/wpt/mozilla.

If anyone has anything to suggest or add, please let me know! Otherwise, I'll continue on and hopefully have a PR ready sometime this weekend.

**Comment by jdm** [comment]
How are you running your tests? You should really only be running a single directory like resiurce-tim bg, not the whole testsuite
